extensions/azurecore/src/account-provider/auths/azureAuth.ts

```diff
@@ -0,0 +1,439 @@
+	 */
+	token: string;
+
+	/**
+	 * TokenType
+	 */
+	tokenType: string;
+}
+
+export interface TokenClaims { // https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens
+	aud: string;
+	iss: string;
+	iat: number;
+	idp: string,
+	nbf: number;
+	exp: number;
+	c_hash: string;
+	at_hash: string;
+	aio: string;
+	preferred_username: string;
+	email: string;
+	name: string;
+	nonce: string;
+	oid: string;
+	roles: string[];
+	rh: string;
+	sub: string;
+	tid: string;
+	unique_name: string;
+	uti: string;
+	ver: string;
+}
+
+export type TokenRefreshResponse = { accessToken: AccessToken, refreshToken: RefreshToken, tokenClaims: TokenClaims };
+
+export abstract class AzureAuth {
+
+	protected readonly WorkSchoolAccountType: string = 'work_school';
+	protected readonly MicrosoftAccountType: string = 'microsoft';
+
+	protected readonly loginEndpointUrl: string;
+	protected readonly commonTenant: string;
+	protected readonly redirectUri: string;
+	protected readonly scopes: string[];
+	protected readonly clientId: string;
+	protected readonly resources: Resource[];
+
+
+	constructor(
+		protected readonly metadata: AzureAccountProviderMetadata,
+		protected readonly tokenCache: SimpleTokenCache,
+		protected readonly context: vscode.ExtensionContext,
+		protected readonly authType: AzureAuthType,
+		public readonly userFriendlyName: string
+	) {
+		this.loginEndpointUrl = this.metadata.settings.host;
+		this.commonTenant = 'common';
+		this.redirectUri = this.metadata.settings.redirectUri;
+		this.clientId = this.metadata.settings.clientId;
+		this.scopes = this.metadata.settings.scopes;
+
+		this.resources = [this.metadata.settings.armResource];
+	}
+
+	public abstract async login(): Promise<AzureAccount | azdata.PromptFailedResult>;
+
+	public abstract async autoOAuthCancelled(): Promise<void>;
+
+
+	public async refreshAccess(account: azdata.Account): Promise<azdata.Account> {
+		const refreshTokens: RefreshToken[] = [];
+		for (const resource of this.resources) {
+			const response = await this.getCachedToken(account.key, resource);
+			if (!response) {
+				account.isStale = true;
+				return account;
+			}
+
+			const refreshToken = response.refreshToken;
+			if (!refreshToken || !refreshToken.key) {
+				account.isStale = true;
+				return account;
+			}
+			refreshTokens.push(refreshToken);
+		}
+
+		try {
+			this.refreshAccessTokens(account, refreshTokens);
+		} catch (ex) {
+			if (ex.message) {
+				vscode.window.showErrorMessage(ex.message);
+			}
+			console.log(ex);
+		}
+		return account;
+	}
+
+
+	public async getSecurityToken(account: azdata.Account, azureResource: azdata.AzureResource): Promise<TokenResponse> {
+		const resource = this.resources.find(s => s.azureResourceId === azureResource);
+		if (!resource) {
+			return undefined;
+		}
+
+		const { accessToken } = await this.getCachedToken(account.key, resource);
+
+		const azureAccount = account as AzureAccount;
+
+		const response: TokenResponse = {};
```

Comment:
Parts of our code need a very weird mapping of tenants and subscriptions to tokens. So that's essentially what this does.